feat: adds new relic integration

jemisonf opened the pull request:
Adds a New Relic service which can report raw metrics to New Relic insights using the metrics api and the New Relic Python SDK. Includes tests and updates to configuration docs but not a local install script or pre-built dashboards.

We have been running this code in a test Spinnaker cluster for ~1.5 weeks now without issues. We've also built the config into halyard in a branch [here](https://github.com/newrelic-forks/spinnaker-halyard/tree/newrelic-metric-store) for use configuring the monitoring daemon.

dotdotdotpaul commented on spinnaker-monitoring-daemon/spinnaker-monitoring/newrelic_service.py:
```diff
@@ -0,0 +1,82 @@
+
+
+class NewRelicMetricsService(object):
+    """A metrics servie for interacting with New Relic."""
```
Typo:  misspelled "service"

dotdotdotpaul commented on spinnaker-monitoring-third-party/third_party/newrelic/install.sh:
```diff
@@ -0,0 +1,2 @@
+echo "Hello world. This is a stub of an install script for New Relic."
+pip install newrelic-telemetry-sdk
```
Is this still a stub?  Or does `pip install` actually do the needful?  😄

jemisonf replied:
it's not required, `newrelic-telemetry-sdk` is in the `requirements.txt` file as well. Not sure how important this script actually is, I could probably just remove it.

robzienert replied:
https://github.com/spinnaker/spinnaker-monitoring/search?q=install.sh&unscoped_q=install.sh

Doesn't look necessary. Let's get rid of it!

robzienert commented on spinnaker-monitoring-daemon/spinnaker-monitoring/__main__.py:
```diff
@@ -157,7 +160,7 @@ def main(config_search_path):
   if not isinstance(stores, list):
     stores = [stores]
-  stores.extend([store for store in ['datadog', 'prometheus', 'stackdriver']
+  stores.extend([store for store in ['datadog', 'prometheus', 'stackdriver', 'newrelic']
```
🤔 Can you pull this out into a constant or something? I don't like that there's a list just floating out there in some method that we have to change.

robzienert commented on spinnaker-monitoring-daemon/spinnaker-monitoring/newrelic_service.py:
```diff
@@ -0,0 +1,82 @@
+        metric_list = []
+        spectator_client.foreach_metric_in_service_map(
+            service_metrics, self.parse_metric, metric_list)
+        # using 1000 as a known-good size, not a theoretical maximum
```
Thanks for the clarification 💯

robzienert commented on spinnaker-monitoring-daemon/spinnaker-monitoring/newrelic_service.py:
```diff
@@ -0,0 +1,82 @@
+    elif 'NEWRELIC_INSERT_KEY' in os.environ:
+        insert_key = os.environ['NEWRELIC_INSERT_KEY']
+    else:
+        raise "New Relic is enabled but the config file has no New Relic Insights Insert Key option"
```
How obvious are insert key errors to New Relic customers? Would it be beneficial to include a permalink to some docs on them when this error is raised?

robzienert commented:
Cool, thank you!

jemisonf commented:
@robzienert thanks for the feedback! Those suggestions all looked good to me. I pulled the metric list into a constant defined at the top of the file (not sure if there's a better python convention for that), removed the third party install script, and added a docs link the exception.

robzienert commented:
Cool. Looks like the most recent commit changed a lot more (a bunch of whitespace changes?) and the tests are breaking. Once things are passing, we'll get this in.

jemisonf commented:
It looks like autopep8 redid the formatting in `__main__.py`. I'll fix that really quick. The build error is . . . something to do with trying to fetch the branch? I'm wondering if squashing the commits to match the preferred commit message style will fix it.

jemisonf commented:
Looks like it's working!

jwelch92 commented:
Pushed another change to add copyright blurbs to the newly added files.

ewiseblatt commented on spinnaker-monitoring-daemon/spinnaker-monitoring/newrelic_service.py:
```diff
@@ -0,0 +1,116 @@
+            tags["applicationName"] = service_data["applicationName"]
+            tags["applicationVersion"] = service_data["applicationVersion"]
+            interval = metric_data_value["t"] - \
+                service_data["__collectStartTime"]
```
I dont think you want this as your interval. collectStartTime is when you started collecting the data. "t" is the time spinnaker recorded the data. I presume the "interval" that the API wants is the time period that the count was accumulated over, which would be the start time of the process providing the metric, service_metadata['startTime'].
So interval might be metric_data_value['t'] - service_metadata['startTime']. Hopefully both are in the same time units, which is probably epoch milliseconds. Otherwise there might be a conversion involved.

jemisonf replied:
Thanks, this looks like a legit bug. Would be happy to open a second PR with fixes for this.

ewiseblatt commented on spinnaker-monitoring-daemon/spinnaker-monitoring/newrelic_service.py:
```diff
@@ -0,0 +1,116 @@
+                  for i in xrange(0, len(metric_list), chunk_size)]
+        for chunk in chunks:
+            response = self.metric_client.send_batch(chunk)
+            response.raise_for_status()
```
I'm not familiar with the API but sounds like this is going to raise an exception on failure.
Is that really what you want? It could be but might not. I'm just asking.

jemisonf replied:
That's the intention, since failure would generally mean either the endpoint or the API key is misconfigured. Let me know if you have a better pattern for handling this in mind.

jemisonf replied:
Here's what that looks like in the logs:
```
12:34:46 INFO Using default metric filter for "igor"
12:34:46 INFO Using default metric transformer for "igor"
12:34:46 DEBUG Collection times 5698 (ms): {'fiat': 4, 'igor': 5665, 'clouddriver': 1, 'rosco': 8, 'kayenta': 19, 'echo': 0, 'front50': 1, 'gate': 17, 'orca': 16}
12:34:46 ERROR Traceback (most recent call last):
  File "spinnaker-monitoring-daemon/spinnaker-monitoring/server_handlers.py", line 205, in __call__
    count = service.publish_metrics(service_metric_map)
  File "spinnaker-monitoring-daemon/spinnaker-monitoring/newrelic_service.py", line 56, in publish_metrics
    response = self.metric_client.send_batch(chunk)
  File "/home/fischer/.local/lib/python2.7/site-packages/newrelic_telemetry_sdk/client.py", line 109, in send_batch
    return self.session.post(self.url, headers=headers, data=payload)
  File "/home/fischer/.local/lib/python2.7/site-packages/requests/sessions.py", line 581, in post
    return self.request('POST', url, data=data, json=json, **kwargs)
  File "/home/fischer/.local/lib/python2.7/site-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/fischer/.local/lib/python2.7/site-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "/home/fischer/.local/lib/python2.7/site-packages/requests/adapters.py", line 516, in send
    raise ConnectionError(e, request=request)
ConnectionError: HTTPSConnectionPool(host='abc.def', port=443): Max retries exceeded with url: /metric/v1 (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f5155817d10>: Failed to establish a new connection: [Errno -2] Name or service not known',))
```

The monitoring daemon continues running after the exception is raised, so in the case of something like New Relic being down this wouldn't require a restart.

ewiseblatt replied:
yea, I see that the daemon catches exceptions that you throw to protect against the case I was worried about so this is fine.

ewiseblatt commented on spinnaker-monitoring-daemon/spinnaker-monitoring/newrelic_service.py:
```diff
@@ -0,0 +1,116 @@
+    def __init__(self, spectator_helper, metric_client, tags, options):
+        self.spectator_helper = spectator_helper
+        self.metric_client = metric_client
+        self.tags = tags
```
might want to call self.tags "self.impliedTags" or something like that.

ewiseblatt commented on spinnaker-monitoring-daemon/spinnaker-monitoring/newrelic_service.py:
```diff
@@ -0,0 +1,116 @@
+            tags = self.tags.copy()
+            for tag in metric_instance["tags"]:
+                tags[tag["key"]] = tag["value"]
+            tags["applicationName"] = service_data["applicationName"]
```
I wouldnt add applicationName here. If you want the applicationName then you can turn on "inject_service_tags" config flag and it will add tags "spin_service" and "spin_variant" where "spin_service" is something like "clouddriver" and "spin_variant" is something like "readonly" for a "clouddriver-readonly" deployment. The variant is empty if it is a standard configuration or not a HA deployment.

Also the metric names encode the service name by default. You can turn that off setting "decorate_metric_name" to false.

ewiseblatt commented on spinnaker-monitoring-daemon/spinnaker-monitoring/newrelic_service.py:
```diff
@@ -0,0 +1,116 @@
+        options)
+    if 'newrelic' in options and 'insert_key' in options['newrelic']:
+        insert_key = options['newrelic']['insert_key']
+    elif 'NEWRELIC_INSERT_KEY' in os.environ:
```
Dont environment variables usually have higher precedence than config options?
Or are these options command-line ?

jemisonf replied:
Yeah, the config file is overwriting environment variables here. Should probably be the opposite.

ewiseblatt commented on spinnaker-monitoring-daemon/tests/newrelic_service_test.py:
```diff
@@ -0,0 +1,287 @@
+def get_gauge_metric():
+    return (
+        {
+            "applicationName": "spinnaker-monitoring",
```
I know this is random test data, but to be clear the applicationName wont be "spinnaker-monitoring". It will be the service name (e.g. "clouddriver").

jemisonf commented on spinnaker-monitoring-daemon/spinnaker-monitoring/newrelic_service.py:
```diff
@@ -0,0 +1,116 @@
+            if kind == spectator_client.GAUGE_PRIMITIVE_KIND:
+                metric_list.append(GaugeMetric(
+                    name=metric_name, value=metric_data_value["v"], tags=tags))
+            else:
```
Possibly making an obvious mistake here but I don't see this defined anywhere in the project and it doesn't look like it's a possible return of `SpectatorClientHelper.determine_primitive_kind`.

ewiseblatt replied:
https://github.com/spinnaker/spinnaker-monitoring/blob/master/spinnaker-monitoring-daemon/spinnaker-monitoring/spectator_client.py#L346

jemisonf replied:
Ah, gotcha. I was going by other parts of code (ie the datadog service) that handle SUMMARY_KIND as a counter, but given that the New Relic SDK has a summary metric it doesn't seem like there's any reason not to use that.
